enhancements/machine-config/manage-boot-images.md

```diff
@@ -282,44 +285,108 @@ spec:
 ```
 #### Tracking boot image history
 
-This proposal will also introduce a new CR, `MachineSetBootImageHistory` for tracking boot image history. As a starting point, here is a stub type definition for this:
+This proposal will also introduce a new CR, `BootImageHistory` for tracking boot image history. As a starting point, here is a stub type definition for this:
 
 ```
-type MachineSetBootImageHistory struct {
+type BootImageHistory struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   MachineSetBootImageHistorySpec   `json:"spec,omitempty"`
-	Status MachineSetBootImageHistoryStatus `json:"status,omitempty"`
+	Spec   BootImageHistorySpec   `json:"spec,omitempty"`
+	Status BootImageHistoryStatus `json:"status,omitempty"`
+}
+
+// BootImageHistorySpec defines the desired state of BootImageHistory
+type BootImageHistorySpec struct {
 }
 
-// MachineSetBootImageHistorySpec defines the desired state of MachineSetBootImageHistory
-type MachineSetBootImageHistorySpec struct {
-	MachineSetName string                   `json:"machineSetName"`
-	Details        []BootImageHistoryDetail `json:"details"`
+// BootImageHistoryStatus defines the observed state of BootImageHistory
+type BootImageHistoryStatus struct {
+	// machineResourceReference contains identifying information of the machine management resource being tracked.
+	// +kubebuilder:validation:Required
+	// +required
+	MachineResourceReference MachineResourceReference `json:"machineResourceReference"`
+	// details is a list of boot image history entries of the machine resource.
+	// +optional
+	Details []BootImageHistoryDetail `json:"details,omitempty"`
 }
 
-// MachineSetBootImageHistoryStatus defines the observed state of MachineSetBootImageHistory
-type MachineSetBootImageHistoryStatus struct {
+type MachineResourceReference struct {
+	// name is the machine management resource's name
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+	// kind is the machine management resource's kind
+	// +kubebuilder:validation:Required
+	// +required
+	Kind string `json:"kind"`
+	// apiGroup is name of the APIGroup that the machine management resource belongs to. This is for disambiguating
+	// between Cluster API and Machine API backed resources.
+	// +kubebuilder:validation:Required
+	// +required
+	APIGroup string `json:"apiGroup"`
 }
 
 // BootImageHistoryDetail is the struct for each element in the Details array
 type BootImageHistoryDetail struct {
```

Comment:
I wonder whether it is worth capturing the resource version of the request that updated to this image? So that you could, in theory, recover the complete state at the time of the update using the resource version?

Reply:
I'm not very familiar with resource versions, so I'll try to investigate this more. When you say complete state in this context, do you mean that of the configmap, or something else? I wasn't aware you could re-construct the configmap like that! If that is indeed the case, we could technically store the resourceversion of the configmap and won't need to store the "boot image ref" which may vary from platform to platform

Reply:
It would probably be more idiomatic to capture the generation of the configmap, though I can't quite remember what I was thinking.
To a degree, you could see history by storing the resource version but it's not guaranteed, its an etcd feature.

Knowing which generation triggered the update would allow you to check that the update controller had seen the latest generation of the configmap, or at least, tie this detail to the particular generation.